### PR TITLE
More efficient cmap compilation

### PR DESCRIPTION
**Pull request by cmyr**
This produces significantly smaller cmap tables by combining segments where possible, and taking advantage of of the `offset_range_id` array to store glyphids directly where this is beneficial.

This fixes the issue where our cmap tables were often significantly larger than the ones produced by the python toolchain.

I believe there may be one further optimization possible here, which i am curious to explore but am holding back because I think the size savings are negligible: but I believe it is theoretically possible to reuse glyph ids in the `glyphIDArray`, if two segments have glyphs identical spacing patterns, due to the fact that the idDelta is still applied even if the glyphidarray is used.

### CODE REVIEW

File: write-fonts/src/tables/cmap.rs [hunk 1]
```diff
@@ -237,6 +236,187 @@ impl Cmap {
...
+        return Default::default();
+    };
+
+    let mut result = vec![first];
+
+    // now we want to collect the segments, combining smaller segments where
+    // that leads to a size savings.
+    //
+    // This differs from the python, which starts from larger segments and then
+    // subdivides them, but the overall idea is the same.
```
Comment (rsheeter) [inline write-fonts/src/tables/cmap.rs:383]: Why do we choose to differ from python?

File: write-fonts/src/tables/cmap.rs [hunk 2]
```diff
@@ -237,6 +236,187 @@ impl Cmap {
...
+        };
+        self.seg_start += seg_len + 1;
+        self.currently_contiguous = false;
+        result
+    }
+}
+
+// this iterator creates segments greedily, that is it will create a new segment
+// at every opportunity. These are then recombined by the caller to generate
+// the most efficient overall sequence of segments.
```
Comment (rsheeter) [inline write-fonts/src/tables/cmap.rs:321]: I'm not quite sure what this is trying to tell me, why this is important. Perhaps we could mention that?

File: write-fonts/src/tables/cmap.rs [hunk 3]
```diff
@@ -237,6 +236,187 @@ impl Cmap {
...
+    id_delta: Option<i32>,
+}
+
+impl Format4Segment {
+    fn len(&self) -> usize {
+        self.end_ix - self.start_ix + 1
+    }
+
+    // cost in bytes of this segment
+    fn cost(&self) -> usize {
```
Comment (rsheeter) [inline write-fonts/src/tables/cmap.rs:260]: add a spec ref?

File: write-fonts/src/tables/cmap.rs [hunk 4]
```diff
@@ -237,6 +236,187 @@ impl Cmap {
...
+        let glyph_id_cost = self
+            .id_delta
+            .is_none()
+            .then(|| self.len() * u16::RAW_BYTE_LEN)
+            .unwrap_or(0);
+        BASE_COST + glyph_id_cost
+    }
+
+    fn combine(&self, other: Format4Segment) -> Format4Segment {
+        assert_eq!(other.start_ix, self.end_ix + 1);
```
Comment (rsheeter) [inline write-fonts/src/tables/cmap.rs:271]: Not loving the assert, I suggest modelling the failure explicitly, e.g. try_combine or similar.

File: write-fonts/src/tables/cmap.rs [hunk 5]
```diff
@@ -237,6 +236,187 @@ impl Cmap {
...
+        let Some(((mut prev_cp, mut prev_gid), rest)) =
+            self.mappings[self.seg_start..].split_first()
+        else {
+            // if this is the last element, make a final segment
+            return Some(self.make_segment(0));
+        };
+
+        for (i, (cp, gid)) in rest.iter().enumerate() {
+            // first: all segments must be a contiguous range of codepoints
+            if *cp as u32 - prev_cp as u32 > 1 {
```
Comment (rsheeter) [inline write-fonts/src/tables/cmap.rs:338]: The check is throwing me, can' t we just check `if cp != prev_cp + 1`, e.g. more directly say "if cp doesn't continue the run"

File: write-fonts/src/tables/cmap.rs (hunk 2 above)
Comment (rsheeter) [reply write-fonts/src/tables/cmap.rs:321]: Actually that sort of describes the code in general. I think it's saying it tries to find runs, prioritizing, from most to least favorable:

1. A run of at least 2 where codepoint *and* gid are contiguous
1. A run of at least 2 where codepoints are contiguous (but gids aren't)
1. A run of 1

I think the comment might want to elaborate.

Comment (rsheeter) [reply write-fonts/src/tables/cmap.rs:321]: https://github.com/googlefonts/fontations/pull/1229 attempts to simplify, eliminating the currently_contiguous field and making the iterator just grab the next run, prioritized as ^, each cycle.

File: write-fonts/src/tables/cmap.rs [hunk 6]
```diff
@@ -237,6 +236,187 @@ impl Cmap {
...
+            continue;
+        }
+
+        // next: if chars are contiguous and neither has a delta, we always combine
+        // this will mostly happen if we've combined the previous, previously
+        // delta-having segment
+        if is_contiguous_with_prev && prev.id_delta.is_none() && segment.id_delta.is_none() {
+            *prev = prev.combine(segment);
+            continue;
+        }
```
Comment (rsheeter) [inline write-fonts/src/tables/cmap.rs:407]: Why isn't the combine if bytes saved check sufficient?

File: write-fonts/src/tables/cmap.rs [hunk 7]
```diff
@@ -237,6 +236,187 @@ impl Cmap {
...
+            let combined = prev.combine(segment);
+            if combined.cost() < prev.cost() + segment.cost() {
+                *prev = combined;
+                continue;
+            }
+        }
+        result.push(segment);
+    }
+    result
+}
```
Comment (rsheeter) [inline write-fonts/src/tables/cmap.rs:419]: I'm slightly surprised the loop doesn't mirror the comment more directly, e.g. see if we can combine, do so if it saves bytes? - naive pseudocode:

```rust
while let Some(next) = iter.next() {
  // we initialized result with the first result from iter
  if !result.last_mut().unwrap().try_combine(next) {
    result.push(next);
  }
}
```

Checking next and prev and checking id delta instead of letting the cost check handle it both throw me.

Comment (rsheeter) [reply write-fonts/src/tables/cmap.rs:419]: https://github.com/googlefonts/fontations/pull/1230 tries ^ and seems to pass the tests.

File: write-fonts/src/tables/cmap.rs (hunk 4 above)
Comment (rsheeter) [reply write-fonts/src/tables/cmap.rs:271]: https://github.com/googlefonts/fontations/pull/1230 explores this idea

**Review by rsheeter — CHANGES_REQUESTED** [review]
Generally looks good but I found some of the code hard to follow. I sent suggestions for simplifications. I'm not tied ot the specific simplifications I proposed but I would like to look again once it's been altered to be a bit easier to follow so marking as Request changes.

File: write-fonts/src/tables/cmap.rs (hunk 2 above)
Comment (cmyr) [reply write-fonts/src/tables/cmap.rs:321]: I'm trying to explain a slightly arbitrary design choice; that this iterator returns segments but that they need further optimization. In my initial design the iterator produced the segments directly, but you need to do some lookahead to determine if you should split a segment or not.

File: write-fonts/src/tables/cmap.rs (hunk 4 above)
Comment (cmyr) [reply write-fonts/src/tables/cmap.rs:271]: this is just documenting an invariant, we only combine contiguous segments.  Can add a comment...

File: write-fonts/src/tables/cmap.rs (hunk 6 above)
Comment (cmyr) [reply write-fonts/src/tables/cmap.rs:407]: good catch, I didn't write the `cost()` check until I'd gotten to that lat branch that has the tricky logic, but you're right that it should cover this case as well.

File: write-fonts/src/tables/cmap.rs (hunk 7 above)
Comment (cmyr) [reply write-fonts/src/tables/cmap.rs:419]: yea I think this is a nice idea, will play around a bit.

File: write-fonts/src/tables/cmap.rs (hunk 3 above)
Comment (cmyr) [reply write-fonts/src/tables/cmap.rs:260]: there's no specific part of the spec that describes a segment, but I'll try to make this clearer..

File: write-fonts/src/tables/cmap.rs [hunk 8]
```diff
@@ -238,6 +237,222 @@ impl Cmap {
...
+            let can_merge_left = prev.can_merge(&current);
+
+            if !can_merge_left {
+                result.push(current);
+                continue;
+            }
+
+            let combined = prev.combine(current);
+            let should_merge = match next.as_ref() {
+                // if we can merge on both sides, consider total cost
```
Comment (rsheeter) [inline write-fonts/src/tables/cmap.rs:438]: Why do you need to check both sides at once? That is, why doesn't checking if current could merge with the last entry in result suffice?

File: write-fonts/src/tables/cmap.rs [hunk 9]
```diff
@@ -238,6 +237,222 @@ impl Cmap {
...
+        self.seg_start += seg_len + 1;
+        self.gids_in_order = false;
+        result
+    }
+
+    /// Find the next possible segment.
+    ///
+    /// A segment _must_ be a contiguous range of chars, but we where such a range
+    /// contains subranges that are also contiguous ranges of glyph ids, we will
+    /// split those subranges into separate segments.
```
Comment (rsheeter) [inline write-fonts/src/tables/cmap.rs:345]: This seems more complicated than just finding the next run more directly? - we've got various special cases that go away if we do that:

1. `if prev gids were ordered but this one isn't, end prev segment`
1. tracking `self.gids_in_order`
1. `if gids were not previously ordered but are now`
1. `done looping then create the last segment`

https://github.com/googlefonts/fontations/pull/1229/files avoids all this. Is there a reason this is better that I'm missing?

File: write-fonts/src/tables/cmap.rs (hunk 8 above)
Comment (rsheeter) [reply write-fonts/src/tables/cmap.rs:438]: Per IM, let's add a comment explaining the arithmatic of merging that leads to potential for merging 3 to be a win where 2 wasn't, and no need to check 4+

**Review by rsheeter — APPROVED** [review]
LGTM subject to remaining comments being addressed
